### PR TITLE
Exempt `__repr__`s, `__slots__` and other `@attrs.define` related changes from compatibility policy.

### PR DESCRIPTION
**Pull request by graingert**
I think it's generally assumed that `__repr__` and monkey patching are not public interfaces to be depended upon, but now with `@attrs.frozen` and `@attrs.define` on the horizon it's likely that objects new and old will be better with the new attrs defaults.

This PR nails down those compatibility assumptions, for the purposes of `@functools.wraps`, `@property` and `@attrs.frozen`. Specifically this PR doesn't materially change the compatibility it just documents some unspoken assumptions previously implied by  "Here is a non-exhaustive list".

For example, looking through the log with `git log -S"def __repr__"` finds a few places where these changes have been made

## Contributor Checklist:

* [x] The associated ticket in Trac is here:https://twistedmatrix.com/trac/ticket/9982
* [ ] I ran `tox -e black-reformat` to format my patch to meet the [Twisted Coding Standard](https://twistedmatrix.com/documents/current/core/development/policy/coding-standard.html)
* [ ] I ran additional checks listed at [Getting Your Patch Accepted](https://twistedmatrix.com/trac/wiki/TwistedDevelopment#GettingYourPatchAccepted)
* [x] I have created a newsfragment in src/twisted/newsfragments/ (see: [News files](https://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles))
* [ ] I have updated the automated tests.
* [x] I have submitted the associated Trac ticket for review by adding the word `review` to the keywords field in Trac, and putting a link to this PR in the comment; it shows up in https://twisted.reviews/ now.

### CODE REVIEW

File: docs/core/development/policy/compatibility-policy.rst [hunk 1]
```diff
@@ -180,6 +180,13 @@ Even though Python objects can be pickled and unpickled without explicit support
 Because of this, no application may depend on any object defined by Twisted to provide pickle compatibility between any release unless the object explicitly documents this as a feature it has.
 
 
+Representations
+^^^^^^^^^^^^^^^
+
+The printable representaions of objects, as returned by ``repr(object)`` and defined by ``def __repr__(self):`` are for debugging and informational purposes.
+Because of this, no application may depend on any object defined by Twisted to provide repr compatability between any release.
```
Comment (graingert) [inline docs/core/development/policy/compatibility-policy.rst]: I also considered a note about how applications should also not rely on `eval(repr(object)) == object` compatability

Comment (mthuurne) [reply docs/core/development/policy/compatibility-policy.rst]: Typos: "compatability" -> "compatibility" and "representaions" -> "representations".

Also, perhaps "applications must not depend" is clearer than "no application may depend".

File: src/twisted/newsfragments/9982.misc [hunk 2]
```diff
@@ -0,0 +1 @@
+Exempt ``__repr__``s from compatability policy.
```
Comment (mthuurne) [inline src/twisted/newsfragments/9982.misc]: There is a `doc` category for documentation changes, that may be more appropriate than `misc` since this is not a trivial change.

Also "compatability" -> "compatibility" here as well.

File: docs/core/development/policy/compatibility-policy.rst [hunk 3]
```diff
@@ -162,22 +162,29 @@ The most basic thing that can happen between Twisted versions, of course, is tha
...
 
 Even though Python objects can be pickled and unpickled without explicit support for this, whether a particular pickled object can be unpickled after any particular change to the implementation of that object is less certain.
-Because of this, no application may depend on any object defined by Twisted to provide pickle compatibility between any release unless the object explicitly documents this as a feature it has.
+Because of this, applications must not depend on any object defined by Twisted to provide pickle compatibility between any release unless the object explicitly documents this as a feature it has.
+
+
+Representations
+^^^^^^^^^^^^^^^
+
+The printable representaions of objects, as returned by ``repr(object)`` and defined by ``def __repr__(self):`` are for debugging and informational purposes.
```
Comment (mthuurne) [inline docs/core/development/policy/compatibility-policy.rst]: Maybe write `repr(obj)` or `repr(...)` instead of `repr(object)` to avoid confusion with the `object` builtin. Although I don't think that kind of confusion is very likely.

**Review by mthuurne — APPROVED** [review]
I suggested some changes; with those resolved I think the patch is fine.

As for the policy change itself, I think that is helpful so we are free to switch to attrs-generated `__repr__()` implementations where possible. But it would be good to get some more opinions on this, since there might be a scenario that I'm not aware of in which repr changes can be a problem.

File: src/twisted/newsfragments/9982.misc (hunk 2 above)
Comment (graingert) [reply src/twisted/newsfragments/9982.misc]:
```suggestion
Exempt ``__repr__``s from compatibility policy.
```

File: docs/core/development/policy/compatibility-policy.rst (hunk 3 above)
Comment (graingert) [reply docs/core/development/policy/compatibility-policy.rst]:
```suggestion
The printable representations of objects, as returned by ``repr(obj, /)`` and defined by ``def __repr__(self):`` are for debugging and informational purposes.
```

Comment (graingert) [reply docs/core/development/policy/compatibility-policy.rst]: I went for the notation from `help(repr)`

**Comment by graingert** [comment]
> I'm not aware of in which repr changes can be a problem.

I think that could only be systems that rely on the `eval(repr(obj)) == obj` invariant

File: docs/core/development/policy/compatibility-policy.rst [hunk 4]
```diff
@@ -162,22 +162,29 @@ The most basic thing that can happen between Twisted versions, of course, is tha
...
 Even though Python objects can be pickled and unpickled without explicit support for this, whether a particular pickled object can be unpickled after any particular change to the implementation of that object is less certain.
-Because of this, no application may depend on any object defined by Twisted to provide pickle compatibility between any release unless the object explicitly documents this as a feature it has.
+Because of this, applications must not depend on any object defined by Twisted to provide pickle compatibility between any release unless the object explicitly documents this as a feature it has.
+
+
+Representations
+^^^^^^^^^^^^^^^
+
+The printable representations of objects, as returned by ``repr(obj, /)`` and defined by ``def __repr__(self):`` are for debugging and informational purposes.
+Because of this, applications must not depend on any object defined by Twisted to provide repr compatibility between any release.
```
Comment (graingert) [inline docs/core/development/policy/compatibility-policy.rst:185]: this also allows for renaming functions and methods and providing an alias.

File: docs/core/development/policy/compatibility-policy.rst [hunk 5]
```diff
@@ -162,22 +162,29 @@ The most basic thing that can happen between Twisted versions, of course, is tha
 That means that no application may ever rely on, for example, the value of any **func_code** object's **co_code** attribute remaining stable, or the **checksum** of a .py file remaining stable.
 
 **Docstrings** may also change at any time.
-No application code may expect any Twisted class, module, or method's __doc__ attribute to remain the same.
+Applications must not depend on any Twisted class, module, or method's ``__doc__`` attribute to remain the same.
```
Comment (graingert) [inline docs/core/development/policy/compatibility-policy.rst]:
```suggestion
Applications must not depend on any Twisted class, module, or method's metadata attributes such as ``__module__``, ``__name__``, ``__qualname__``, ``__annotations__`` and ``__doc__`` to remain the same.
```

https://github.com/python/cpython/blob/3.8/Lib/functools.py#L31

**Comment by mthuurne** [comment]
> > I'm not aware of in which repr changes can be a problem.
> 
> I think that could only be systems that rely on the `eval(repr(obj)) == obj` invariant

Maybe in some kind of homebrew pickle alternative? But pickle has been around forever and `eval()` certainly doesn't fix the security risks that pickle has when data comes from untrusted sources.

File: docs/core/development/policy/compatibility-policy.rst (hunk 5 above)
Comment (mthuurne) [reply docs/core/development/policy/compatibility-policy.rst]: Eh, `__module__`, `__name__` and `__qualname__` should be stable across releases, I think.

I agree with `__annotations__`: the current annotations are rough first estimates and it's likely a significant number of them will be updated as Twisted itself and projects using Twisted increase their static type checking coverage.

File: docs/core/development/policy/compatibility-policy.rst [hunk 6]
```diff
@@ -162,22 +162,29 @@ The most basic thing that can happen between Twisted versions, of course, is tha
...
 Even though Python objects can be pickled and unpickled without explicit support for this, whether a particular pickled object can be unpickled after any particular change to the implementation of that object is less certain.
-Because of this, no application may depend on any object defined by Twisted to provide pickle compatibility between any release unless the object explicitly documents this as a feature it has.
+Because of this, applications must not depend on any object defined by Twisted to provide pickle compatibility between any release unless the object explicitly documents this as a feature it has.
+
+
+Representations
+^^^^^^^^^^^^^^^
+
+The printable representations of objects, as returned by ``repr(obj, /)`` and defined by ``def __repr__(self):`` are for debugging and informational purposes.
+Because of this, applications must not depend on any object defined by Twisted to provide repr compatibility between any release.
```
Comment (graingert) [inline docs/core/development/policy/compatibility-policy.rst:190]:
```suggestion
descriptors, slots and weakrefs
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
How an object's attributes are defined and accessed is considered an implementation detail.
To allow backwards compatibility and attribute may be moved from the instance ``__dict__`` into
an ``@property`` or other descriptor based accessor.

Assigning attributes to a constructed object, or monkey patching, is not considered a public use. This is to allow creating frozen and/or slotted classes. However if an object is weak referencable, an application may relay on it to remain weak referencable
Because of this, applications must not depend on any object defined by Twisted to provide ``__dict__`` or ``__slots__`` compatibility between any release.
```

File: docs/core/development/policy/compatibility-policy.rst (hunk 5 above)
Comment (graingert) [reply docs/core/development/policy/compatibility-policy.rst]: > Eh, `__module__`, `__name__` and `__qualname__` should be stable across releases, I think.

That would disallow moving a function or class to another module, and aliasing it for compatibility purposes

File: src/twisted/newsfragments/9982.misc [hunk 7]
```diff
@@ -0,0 +1 @@
+Exempt ``__repr__``s from compatibility policy.
```
Comment (graingert) [inline src/twisted/newsfragments/9982.misc]:
```suggestion
Exempt ``__repr__``s. ``__slots__`` and other ``@attrs.define`` related changes from compatibility policy.
```

File: docs/core/development/policy/compatibility-policy.rst [hunk 8]
```diff
@@ -162,24 +162,38 @@ The most basic thing that can happen between Twisted versions, of course, is tha
...
 
+Representations
+^^^^^^^^^^^^^^^
+
+The printable representations of objects, as returned by ``repr(obj, /)`` and defined by ``def __repr__(self):`` are for debugging and informational purposes.
+Because of this, applications must not depend on any object defined by Twisted to provide repr compatibility between any release.
+descriptors, slots and weakrefs
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+How an object's attributes are defined and accessed is considered an implementation detail.
+To allow backwards compatibility and attribute may be moved from the instance ``__dict__`` into
```
Comment (graingert) [inline docs/core/development/policy/compatibility-policy.rst]:
```suggestion
To allow backwards compatibility an attribute may be moved from the instance ``__dict__`` into
```

File: docs/core/development/policy/compatibility-policy.rst [hunk 9]
```diff
@@ -162,24 +162,38 @@ The most basic thing that can happen between Twisted versions, of course, is tha
...
+
+The printable representations of objects, as returned by ``repr(obj, /)`` and defined by ``def __repr__(self):`` are for debugging and informational purposes.
+Because of this, applications must not depend on any object defined by Twisted to provide repr compatibility between any release.
+descriptors, slots and weakrefs
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+How an object's attributes are defined and accessed is considered an implementation detail.
+To allow backwards compatibility an attribute may be moved from the instance ``__dict__`` into
+an ``@property`` or other descriptor based accessor.
+
+Assigning attributes to a constructed object, or monkey patching, is not considered a public use. This is to allow creating frozen and/or slotted classes. However if an object is weak referencable, an application may relay on it to remain weak referencable
```
Comment (graingert) [inline docs/core/development/policy/compatibility-policy.rst]:
```suggestion
Assigning attributes to a constructed object, or monkey patching, is not considered a public use. This is to allow creating frozen and/or slotted classes. However if an object is weak referencable, an application may depend on it remaining weak referencable
```

File: src/twisted/newsfragments/9982.misc (hunk 7 above)
Comment (mthuurne) [reply src/twisted/newsfragments/9982.misc]: `@attrs` sits at a weird point between `attrs` and `@attr.s`; I'd prefer either of those two more literal terms.

Comment (mthuurne) [reply src/twisted/newsfragments/9982.misc]: Also there seems to be a comma missing in the enumeration.

File: docs/core/development/policy/compatibility-policy.rst (hunk 3 above)
Comment (mthuurne) [reply docs/core/development/policy/compatibility-policy.rst]: The `repr(<something>)` phrase is an example of a call, while `/` is only valid as part of a definition, so I don't think it makes sense to use it here.

File: src/twisted/newsfragments/9982.misc (hunk 7 above)
Comment (graingert) [reply src/twisted/newsfragments/9982.misc]: I do mean the new `@attrs.frozen` interface

File: docs/core/development/policy/compatibility-policy.rst (hunk 5 above)
Comment (mthuurne) [reply docs/core/development/policy/compatibility-policy.rst]: Ah, that makes sense. There is a lot of aliasing in Twisted already to export things from private modules, so those metadata fields are probably not very predictable anyway.

File: docs/core/development/policy/compatibility-policy.rst [hunk 10]
```diff
@@ -162,24 +162,38 @@ The most basic thing that can happen between Twisted versions, of course, is tha
...
 
 Even though Python objects can be pickled and unpickled without explicit support for this, whether a particular pickled object can be unpickled after any particular change to the implementation of that object is less certain.
-Because of this, no application may depend on any object defined by Twisted to provide pickle compatibility between any release unless the object explicitly documents this as a feature it has.
+Because of this, applications must not depend on any object defined by Twisted to provide pickle compatibility between any release unless the object explicitly documents this as a feature it has.
 
 
+Representations
+^^^^^^^^^^^^^^^
+
+The printable representations of objects, as returned by ``repr(obj, /)`` and defined by ``def __repr__(self):`` are for debugging and informational purposes.
```
Comment (graingert) [inline docs/core/development/policy/compatibility-policy.rst]:
```suggestion
The printable representations of objects, as returned by ``repr(<object>)`` and defined by ``def __repr__(self):`` are for debugging and informational purposes.
```

File: docs/core/development/policy/compatibility-policy.rst [hunk 11]
```diff
@@ -162,24 +162,38 @@ The most basic thing that can happen between Twisted versions, of course, is tha
...
+Because of this, applications must not depend on any object defined by Twisted to provide repr compatibility between any release.
+descriptors, slots and weakrefs
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+How an object's attributes are defined and accessed is considered an implementation detail.
+To allow backwards compatibility an attribute may be moved from the instance ``__dict__`` into
+an ``@property`` or other descriptor based accessor.
+
+Assigning attributes to a constructed object, or monkey patching, is not considered a public use. This is to allow creating frozen and/or slotted classes. However if an object is weak referencable, an application may depend on it remaining weak referencable
+Because of this, applications must not depend on any object defined by Twisted to provide ``__dict__`` or ``__slots__`` compatibility between any release.
```
Comment (mthuurne) [inline docs/core/development/policy/compatibility-policy.rst:192]: The note about being weak referencable seems to be out of place here. Also I'm a bit worried about the implications of this: if a class is only weak referencable by accident and not by design, having that restriction on it could cause significant difficulties when making future changes to it.

File: docs/core/development/policy/compatibility-policy.rst [hunk 12]
```diff
@@ -162,24 +162,38 @@ The most basic thing that can happen between Twisted versions, of course, is tha
...
+
+The printable representations of objects, as returned by ``repr(obj, /)`` and defined by ``def __repr__(self):`` are for debugging and informational purposes.
+Because of this, applications must not depend on any object defined by Twisted to provide repr compatibility between any release.
+descriptors, slots and weakrefs
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+How an object's attributes are defined and accessed is considered an implementation detail.
+To allow backwards compatibility an attribute may be moved from the instance ``__dict__`` into
+an ``@property`` or other descriptor based accessor.
+
+Assigning attributes to a constructed object, or monkey patching, is not considered a public use. This is to allow creating frozen and/or slotted classes. However if an object is weak referencable, an application may depend on it remaining weak referencable
```
Comment (mthuurne) [inline docs/core/development/policy/compatibility-policy.rst]: There are two types of assigning though: you can add new attributes or update existing attributes. For converting to slotted classes, only adding new attributes has to be forbidden.

While I like immutable objects in general, I don't think Twisted's design is prepared for that. I've had to update attributes after construction in multiple cases. While some of those might have been from improper API use or bad code structure, I don't think that covers all of them.

File: docs/core/development/policy/compatibility-policy.rst (hunk 11 above)
Comment (graingert) [reply docs/core/development/policy/compatibility-policy.rst:192]: Yeah I'm not sure where to put it. Probably in the "Changes Covered by the Compatibility Policy".

>  if a class is only weak referencable by accident and not by design, having that restriction on it could cause significant difficulties when making future changes to it.

It's basically a footgun in `__slots__`, I don't think there's a good reason to make something not weak-referencable, and you can't do it in pypy anyway.

> Without a __weakref__ variable for each instance, classes defining __slots__ do not support weak references to its instances. If weak reference support is needed, then add '__weakref__' to the sequence of strings in the __slots__ declaration.

more: https://github.com/python-attrs/attrs/issues/174

File: docs/core/development/policy/compatibility-policy.rst (hunk 12 above)
Comment (graingert) [reply docs/core/development/policy/compatibility-policy.rst]: > I've had to update attributes after construction in multiple cases. 

updating an attribute is fine, but adding new ones that Twisted never defined seems useless?

Comment (graingert) [reply docs/core/development/policy/compatibility-policy.rst]: going 'round freezing objects willy nilly seems a bit dangerous. But making everything slotted seems safe. Adding an attribute to an object that twisted is unaware of cannot be a public use of that object

Comment (mthuurne) [reply docs/core/development/policy/compatibility-policy.rst]: I think we agree, but the text itself needs updating:
- applications adding new attributes is not supported, so Twisted can move to slotted classes
- freezing attributes should be handled on a case-by-case basis and will likely require deprecation of a setter property in many cases

File: docs/core/development/policy/compatibility-policy.rst (hunk 11 above)
Comment (mthuurne) [reply docs/core/development/policy/compatibility-policy.rst:192]: After discussion on IRC, it seems I misunderstood what it means to be weak referencable. So it might be worthwhile to define that term here, or link to a definition elsewhere if it already exists.

File: src/twisted/newsfragments/9982.doc [hunk 13]
```diff
@@ -0,0 +1 @@
+Exempt ``__repr__``s. ``__slots__`` and other ``@attrs.define`` related changes from compatibility policy.
```
Comment (mthuurne) [inline src/twisted/newsfragments/9982.doc]:
```suggestion
Exempt ``__repr__``, ``__slots__`` and other ``@attrs.define`` related changes from compatibility policy.
```

File: docs/core/development/policy/compatibility-policy.rst [hunk 14]
```diff
@@ -162,24 +162,38 @@ The most basic thing that can happen between Twisted versions, of course, is tha
...
+The printable representations of objects, as returned by ``repr(<object>)`` and defined by ``def __repr__(self):`` are for debugging and informational purposes.
+Because of this, applications must not depend on any object defined by Twisted to provide repr compatibility between any release.
+descriptors, slots and weakrefs
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+How an object's attributes are defined and accessed is considered an implementation detail.
+To allow backwards compatibility an attribute may be moved from the instance ``__dict__`` into
+an ``@property`` or other descriptor based accessor.
+
+Assigning attributes to a constructed object, or monkey patching, is not considered a public use. This is to allow creating frozen and/or slotted classes. However if an object is weak referencable, an application may depend on it remaining weak referencable
+Because of this, applications must not depend on any object defined by Twisted to provide ``__dict__`` or ``__slots__`` compatibility between any release.
```
Comment (graingert) [inline docs/core/development/policy/compatibility-policy.rst]:
```suggestion
Adding new attributes to a constructed object, or monkey patching, is not considered a public use. This is to allow creating slotted classes.
Because of this, applications must not depend on any object defined by Twisted to provide ``__dict__`` or ``__slots__`` compatibility between any release.
```

just going to remove the bit about `__weakref__` if we can agree that all objects that could be weak reference-able should be.

File: docs/core/development/policy/compatibility-policy.rst (hunk 12 above)
Comment (wsanchez) [reply docs/core/development/policy/compatibility-policy.rst]: Careful.  There are cases, I think, where people do depend on being able to attach new attributes on instances of Twisted classes.

In Klein, for example, I have seen (and written) code that attributes a user object to the passed-in `request` object (which comes from `twisted.web`), so you can hold onto something through a request cycle.  Whether that was meant to be allowed or not, breaking this now will create some non-trivial work (perhaps re-architecting stuff) for clients.

Comment (wsanchez) [reply docs/core/development/policy/compatibility-policy.rst]: FWIW `mypy` certainly discourages use of YOLO attributes (I think by default), so once we have everything type hinted, that would be a tool one could use to at least identify some of these use cases upfront.

Comment (graingert) [reply docs/core/development/policy/compatibility-policy.rst]: @wsanchez I think twisted.web.request is ripe for being an @attrs.frozen class :D

Comment (graingert) [reply docs/core/development/policy/compatibility-policy.rst]: @wsanchez klein seems to use request.setComponent, which is explicit support for runtime instance extension

File: docs/core/development/policy/compatibility-policy.rst [hunk 15]
```diff
@@ -162,24 +162,38 @@ The most basic thing that can happen between Twisted versions, of course, is tha
...
-Because of this, no application may depend on any object defined by Twisted to provide pickle compatibility between any release unless the object explicitly documents this as a feature it has.
+Because of this, applications must not depend on any object defined by Twisted to provide pickle compatibility between any release unless the object explicitly documents this as a feature it has.
 
 
+Representations
+^^^^^^^^^^^^^^^
+
+The printable representations of objects, as returned by ``repr(<object>)`` and defined by ``def __repr__(self):`` are for debugging and informational purposes.
+Because of this, applications must not depend on any object defined by Twisted to provide repr compatibility between any release.
+descriptors, slots and weakrefs
```
Comment (altendky) [inline docs/core/development/policy/compatibility-policy.rst]:
```suggestion
Descriptors, Slots and Weak References
```

File: docs/core/development/policy/compatibility-policy.rst [hunk 16]
```diff
@@ -162,24 +162,38 @@ The most basic thing that can happen between Twisted versions, of course, is tha
...
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+How an object's attributes are defined and accessed is considered an implementation detail.
+To allow backwards compatibility an attribute may be moved from the instance ``__dict__`` into
+an ``@property`` or other descriptor based accessor.
```
Comment (altendky) [inline docs/core/development/policy/compatibility-policy.rst]:
```suggestion
To allow backwards compatibility, an attribute may be moved from the instance ``__dict__`` into an ``@property`` or other descriptor based accessor.
```
Semantic newlines, etc etc (and an extra `,`)

File: docs/core/development/policy/compatibility-policy.rst [hunk 17]
```diff
@@ -162,24 +162,38 @@ The most basic thing that can happen between Twisted versions, of course, is tha
...
+
+Adding new attributes to a constructed object, or monkey patching, is not considered a public use. This is to allow creating slotted classes.
```
Comment (altendky) [inline docs/core/development/policy/compatibility-policy.rst]:
```suggestion
Adding new attributes to a constructed object, or monkey patching, is not considered a public use. This restriction allows both creating and converting to slotted classes.
```

**Review by altendky — COMMENTED** [review]
I don't feel I have any authority to make _policy_ changes but I also don't find this to be against any spirit of the policy.  It does seem good to make these extra bits explicit.

**Comment by graingert** [comment]
>  it would be good to get some more opinions on this, since there might be a scenario that I'm not aware of in which repr changes can be a problem.

Unless anyone objects I'm going to merge this PR in 14 days (Friday, October 30, 2020 00:00 UTC)

File: docs/core/development/policy/compatibility-policy.rst (hunk 12 above)
Comment (wsanchez) [reply docs/core/development/policy/compatibility-policy.rst]: `twisted.web.request` is terrible. Try creating an `IRequest` and then describe how much fun that was. My goal in life is for Klein to stop using it but I've been thwarted by year-long review cycles and then I found type hints and other nonsense to gain more traction.

File: docs/core/development/policy/compatibility-policy.rst [hunk 18]
```diff
@@ -97,13 +97,13 @@ Procedure for Exceptions to this Policy
 
 **Every change is unique.**
 
-Sometimes, we'll want to make a change that fits with this spirit of this document (keeping Twisted working for applications which rely upon it) but may not fit with the letter of the procedure described above (the change modifies behavior of an existing API sufficiently that something might break).
-Generally, the reason that one would want to do this is to give applications a performance enhancement or bug fix that could break behavior that unanticipated, hypothetical uses of an existing API, but we don't want well-behaved applications to pay the penalty of a deprecation/adopt-a-new-API/removal cycle in order to get the benefits of the improvement if they don't need to.
```
Comment (wsanchez) [inline docs/core/development/policy/compatibility-policy.rst:101]: I don't this this is more accurate

File: docs/core/development/policy/compatibility-policy.rst [hunk 19]
```diff
@@ -132,8 +132,6 @@ By doing so, there's no chance anything could access these objects by going thro
 
 Test code and test helpers are considered private API and should not be imported outside
 of the Twisted testing infrastructure.
-As an exception to this, :api:`twisted.test.proto_helpers` is considered a public API
-(see `#6435 <https://twistedmatrix.com/trac/ticket/6435>`_ for more discussion).
```
Comment (wsanchez) [inline docs/core/development/policy/compatibility-policy.rst:136]: Did this thinking on this change? Seems unrelated to the topic…

File: docs/core/development/policy/compatibility-policy.rst (hunk 11 above)
Comment (wsanchez) [reply docs/core/development/policy/compatibility-policy.rst:192]: If the reference to weak references (heh) is removed, the title here should lose it also.

File: docs/core/development/policy/compatibility-policy.rst (hunk 4 above)
Comment (wsanchez) [reply docs/core/development/policy/compatibility-policy.rst:185]: Hrm.  I think there are cases where changing the result of `str(obj)` is problematic…. That said I agree that it's usually not the case.

File: docs/core/development/policy/compatibility-policy.rst (hunk 18 above)
Comment (graingert) [reply docs/core/development/policy/compatibility-policy.rst:101]: ?

File: docs/core/development/policy/compatibility-policy.rst (hunk 19 above)
Comment (graingert) [reply docs/core/development/policy/compatibility-policy.rst:136]: @wsanchez https://freenode.logbot.info/twisted-dev/20201015#c5467846

File: docs/core/development/policy/compatibility-policy.rst (hunk 18 above)
Comment (graingert) [reply docs/core/development/policy/compatibility-policy.rst:101]: @wsanchez ?

File: docs/core/development/policy/compatibility-policy.rst [hunk 20]
```diff
@@ -162,24 +160,37 @@ The most basic thing that can happen between Twisted versions, of course, is tha
...
+Because of this, applications must not depend on any object defined by Twisted to provide pickle compatibility between any release unless the object explicitly documents this as a feature it has.
 
 
+Representations
+^^^^^^^^^^^^^^^
+
+The printable representations of objects, as returned by ``repr(<object>)`` and defined by ``def __repr__(self):`` are for debugging and informational purposes.
+Because of this, applications must not depend on any object defined by Twisted to provide repr compatibility between any release.
+Descriptors, Slots and Weak References
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
Comment (graingert) [inline docs/core/development/policy/compatibility-policy.rst]:
```suggestion
Descriptors and Slots
^^^^^^^^^^^^^^^^^^^^^
```

Comment (graingert) [reply docs/core/development/policy/compatibility-policy.rst]:
```suggestion
Attribute Access
^^^^^^^^^^^^^^^^
```

File: docs/core/development/policy/compatibility-policy.rst (hunk 18 above)
Comment (wsanchez) [reply docs/core/development/policy/compatibility-policy.rst:101]: I don't think that `in undesirable` is more accurate than `that unanticipated,`, so I'm not sure that change is appropriate.  "undesirable" is highly subjective, for starters.

Perhaps "unintended" is better?

File: docs/core/development/policy/compatibility-policy.rst (hunk 19 above)
Comment (wsanchez) [reply docs/core/development/policy/compatibility-policy.rst:136]: okie dokie

**Review by wsanchez — CHANGES_REQUESTED** [review]
Let's lose "undesirable"

File: docs/core/development/policy/compatibility-policy.rst [hunk 21]
```diff
@@ -97,13 +97,13 @@ Procedure for Exceptions to this Policy
...
-Generally, the reason that one would want to do this is to give applications a performance enhancement or bug fix that could break behavior that unanticipated, hypothetical uses of an existing API, but we don't want well-behaved applications to pay the penalty of a deprecation/adopt-a-new-API/removal cycle in order to get the benefits of the improvement if they don't need to.
+Sometimes, we'll want to make a change that fits with the spirit of this document (keeping Twisted working for applications which rely upon it) but may not fit with the letter of the procedure described above (the change modifies behavior of an existing API sufficiently that something might break).
+Generally, the reason that one would want to do this is to give applications a performance enhancement or bug fix that could break behavior in undesirable hypothetical uses of an existing API, but we don't want well-behaved applications to pay the penalty of a deprecation/adopt-a-new-API/removal cycle in order to get the benefits of the improvement if they don't need to.
```
Comment (graingert) [inline docs/core/development/policy/compatibility-policy.rst]:
```suggestion
Generally, the reason that one would want to do this is to give applications a performance enhancement or bug fix that could break behavior in unintended hypothetical uses of an existing API, but we don't want well-behaved applications to pay the penalty of a deprecation/adopt-a-new-API/removal cycle in order to get the benefits of the improvement if they don't need to.
```

**Review by wsanchez — APPROVED** [review]
